File: Binary_Search /javascript/2110.js
const input = require('fs').readFileSync('/dev/stdin').toString().split("\n");
const [n,c] = input[0].split(' ').map(Number)
const houses = input.slice(1,input.length-1).map(Number).sort((a,b)=>a-b);
let [minDis,maxDis] = [1,houses[houses.length-1]];

const binarySearch = (minDis,maxDis,target) => {
  if (minDis > maxDis) return 0
  const midDis = minDis + Math.floor((maxDis-minDis)/2);
  let cnt = 1;
  let pos = houses[0];
  let result = 0;
  houses.reduce((acc,cur) => {
    if (cur >= pos+midDis){
      pos = cur;
      cnt++;
    }
  })
  if (cnt >= target) {
    result = midDis;
    result = Math.max(result,binarySearch(midDis+1,maxDis,target));
  }
  else result = binarySearch(minDis,midDis-1,target);
  return result;
}
console.log(binarySearch(minDis,maxDis,c));

File: Binary_Search /python/2110.py
import sys
input = sys.stdin.readline
n,c = map(int,input().split())
arr = sorted([int(input()) for _ in range(n)])

def search(arr,min_distance,max_distance,target):
    if min_distance>max_distance:
        return 0
    result = 0
    while max_distance>= min_distance:
        mid_distance = min_distance + (max_distance-min_distance)//2
        now_pos = arr[0]
        cnt = 1
        for i in range(1,len(arr)):
            if arr[i] >= now_pos + mid_distance:
                cnt += 1
                now_pos = arr[i]
        if cnt>=target:
            result = mid_distance
            min_distance = mid_distance+1
        else:
            max_distance = mid_distance-1
    return result
        
answer = search(arr,1,arr[-1]-arr[0],c)
print(answer)